common/kvstore/tablestore/table_store_builder.go

```diff
@@ -0,0 +1,402 @@
+const metadataSchemaVersionKey = "schema_version"
+
+// The metadataDeletionKey holds the value of the table name that is being deleted if one is currently being deleted.
+// This is required for atomic deletion of tables.
+const metadataDeletionKey = "deletion"
+
+// The current schema version of the metadata table.
+const currentSchemaVersion uint64 = 0
+
+// StoreType describes the underlying store implementation.
+type StoreType int
+
+const (
+	// LevelDB is a LevelDB-backed store.
+	LevelDB StoreType = iota
+	// MapStore is an in-memory store. This store does not preserve data across restarts.
+	MapStore
+)
+
+// Start creates a new TableStore instance of the given type. The store will be created at the given path.
+// This method can be used to instantiate a new store or to load an existing store.
+// This method will set up a table for each table name provided, and will drop all tables not in the list.
+// Dropping a table is irreversible and will delete all data in the table, so be very careful not to call
+// this method with table names omitted by mistake.
+func (t StoreType) Start(logger logging.Logger, path string, tables ...string) (kvstore.TableStore, error) {
+	base, err := buildBaseStore(t, logger, path)
+	if err != nil {
+		return nil, fmt.Errorf("error building base store: %w", err)
+	}
+
+	return start(logger, base, true, tables...)
+}
+
+// Load loads a table store from disk without modifying the table schema. If there is no existing store at the given
+// path, this method will create one and return a store without any tables.
+func (t StoreType) Load(logger logging.Logger, path string) (kvstore.TableStore, error) {
+	base, err := buildBaseStore(t, logger, path)
+	if err != nil {
+		return nil, fmt.Errorf("error building base store: %w", err)
+	}
+
+	return start(logger, base, false)
+}
+
+// Future work: if we ever decide to permit third parties to provide custom store implementations not in this module,
+// we will need to make this method public.
+
+// start creates a new TableStore instance with the given base store and table names. If modifySchema is true, the
+// tables in the store are made to match the given list of tables by adding and removing tables as needed. If
+// modifySchema is false, the tables are loaded as is, and any tables in the provided list are ignored.
+func start(
+	logger logging.Logger,
+	base kvstore.Store,
+	modifySchema bool,
+	tables ...string) (kvstore.TableStore, error) {
+
+	metadataTable := newTableView(base, "metadata", metadataTableID)
+	namespaceTable := newTableView(base, "namespace", namespaceTableID)
+
+	err := validateSchema(metadataTable)
+	if err != nil {
+		return nil, fmt.Errorf("error validating schema: %w", err)
+	}
+
+	tableIDMap, err := loadNamespaceTable(namespaceTable)
+	if err != nil {
+		return nil, fmt.Errorf("error loading namespace table: %w", err)
+	}
+
+	err = handleIncompleteDeletion(logger, base, metadataTable, namespaceTable)
```

Comment:
Shouldn't this deletion happen before even loading the namespace table? Otherwise a name/ID that's still pending for deletion may get reused

Reply:
You are correct that the partially deleted table may already be inside the namespace table. This edge case is handled by `handleIncompleteDeletion()`, which will perform the necessary deletion within the namespace table (last line of `handleIncompleteDeletion()` calls `dropTable()`, which deletes the entry from the namespace table).

Reply:
If there is a pending deletion, the `loadNamespaceTable` will include that deleted table in the returned `tableIDMap` (because that table may be still in the namespace table, haven't fully deleted), which is not desired

Reply:
I walked this code with a debugger. It turns out this was accidentally working, since the method `addAndRemoveTables()` deletes the spurious table ID. There is a unit test that is sensitive to spurious entries in the table ID map, and that was not failing. 

In order to make this more understandable and less easy to accidentally break, I've moved the loading of the table ID map until after `handleIncompleteDeletion()`.